File: go/pkg/kem/kem_test.go
package kem

import (
	"testing"
)

func TestKEM(t *testing.T) {
	// Generate a key pair
	keyPair, err := Keygen()
	if err != nil {
		t.Fatalf("Failed to generate key pair: %v", err)
	}

	// Encapsulate a shared secret
	ciphertext, sharedSecret1, err := Encapsulate(&keyPair.PublicKey)
	if err != nil {
		t.Fatalf("Failed to encapsulate: %v", err)
	}

	// Decapsulate the shared secret
	sharedSecret2, err := Decapsulate(&keyPair.PrivateKey, ciphertext)
	if err != nil {
		t.Fatalf("Failed to decapsulate: %v", err)
	}

	// Verify that the shared secrets match
	if string(sharedSecret1.Bytes()) != string(sharedSecret2.Bytes()) {
		t.Errorf("Shared secrets do not match")
	}
}

func TestCiphertextSerialization(t *testing.T) {
	// Generate a key pair
	keyPair, err := Keygen()
	if err != nil {
		t.Fatalf("Failed to generate key pair: %v", err)
	}

	// Encapsulate a shared secret
	ciphertext, _, err := Encapsulate(&keyPair.PublicKey)
	if err != nil {
		t.Fatalf("Failed to encapsulate: %v", err)
	}

	// Test byte serialization
	bytes := ciphertext.Bytes()
	ciphertext2, err := CiphertextFromBytes(bytes)
	if err != nil {
		t.Fatalf("Failed to deserialize ciphertext: %v", err)
	}

	if string(ciphertext.Bytes()) != string(ciphertext2.Bytes()) {
		t.Errorf("Ciphertext serialization failed")
	}

	// Test hex serialization
	hexStr := ciphertext.String()
	ciphertext3, err := CiphertextFromHex(hexStr)
	if err != nil {
		t.Fatalf("Failed to deserialize ciphertext from hex: %v", err)
	}

	if string(ciphertext.Bytes()) != string(ciphertext3.Bytes()) {
		t.Errorf("Ciphertext hex serialization failed")
	}
}

func TestSharedSecretSerialization(t *testing.T) {
	// Generate a key pair
	keyPair, err := Keygen()
	if err != nil {
		t.Fatalf("Failed to generate key pair: %v", err)
	}

	// Encapsulate a shared secret
	_, sharedSecret, err := Encapsulate(&keyPair.PublicKey)
	if err != nil {
		t.Fatalf("Failed to encapsulate: %v", err)
	}

	// Test hex serialization
	hexStr := sharedSecret.String()
	sharedSecret2, err := SharedSecretFromHex(hexStr)
	if err != nil {
		t.Fatalf("Failed to deserialize shared secret from hex: %v", err)
	}

	if string(sharedSecret.Bytes()) != string(sharedSecret2.Bytes()) {
		t.Errorf("Shared secret hex serialization failed")
	}
}

func TestInvalidCiphertext(t *testing.T) {
	// Test invalid ciphertext length
	_, err := CiphertextFromBytes([]byte{1, 2, 3})
	if err == nil {
		t.Errorf("Expected error for invalid ciphertext length")
	}

	// Test invalid hex string
	_, err = CiphertextFromHex("invalid")
	if err == nil {
		t.Errorf("Expected error for invalid hex string")
	}
}

func TestInvalidSharedSecret(t *testing.T) {
	// Test invalid shared secret length
	_, err := NewSharedSecret([]byte{1, 2, 3})
	if err == nil {
		t.Errorf("Expected error for invalid shared secret length")
	}

	// Test invalid hex string
	_, err = SharedSecretFromHex("invalid")
	if err == nil {
		t.Errorf("Expected error for invalid hex string")
	}
}

File: js/src/kem.test.ts
import { Ciphertext, SharedSecret, keygen, encapsulate, decapsulate } from './kem';
import { describe, test, expect } from '@jest/globals';

describe('KEM', () => {
  describe('Ciphertext', () => {
    test('should create a ciphertext and convert to/from bytes', () => {
      // Generate a key pair
      const keypair = keygen();
      
      // Encapsulate a shared secret
      const { ciphertext } = encapsulate(keypair.publicKey);
      
      // Convert to bytes and back
      const bytes = ciphertext.getBytes();
      const ciphertext2 = Ciphertext.fromBytes(bytes);
      
      // Compare the bytes
      expect(ciphertext2.getBytes()).toEqual(bytes);
    });
    
    test('should convert ciphertext to/from hex', () => {
      // Generate a key pair
      const keypair = keygen();
      
      // Encapsulate a shared secret
      const { ciphertext } = encapsulate(keypair.publicKey);
      
      // Convert to hex and back
      const hex = ciphertext.toHex();
      const ciphertext2 = Ciphertext.fromHex(hex);
      
      // Compare the bytes
      expect(ciphertext2.getBytes()).toEqual(ciphertext.getBytes());
    });
    
    test('should throw error for invalid bytes length', () => {
      expect(() => {
        Ciphertext.fromBytes(new Uint8Array(10));
      }).toThrow('Invalid ciphertext length');
    });
    
    test('should throw error for invalid hex length', () => {
      expect(() => {
        Ciphertext.fromHex('abcd');
      }).toThrow('Invalid hex string length');
    });
  });
  
  describe('SharedSecret', () => {
    test('should create a shared secret and convert to/from bytes', () => {
      // Generate a key pair
      const keypair = keygen();
      
      // Encapsulate a shared secret
      const { sharedSecret } = encapsulate(keypair.publicKey);
      
      // Convert to bytes and back
      const bytes = sharedSecret.getBytes();
      const sharedSecret2 = new SharedSecret(bytes);
      
      // Compare the bytes
      expect(sharedSecret2.getBytes()).toEqual(bytes);
    });
    
    test('should convert shared secret to/from hex', () => {
      // Generate a key pair
      const keypair = keygen();
      
      // Encapsulate a shared secret
      const { sharedSecret } = encapsulate(keypair.publicKey);
      
      // Convert to hex and back
      const hex = sharedSecret.toHex();
      const sharedSecret2 = SharedSecret.fromHex(hex);
      
      // Compare the bytes
      expect(sharedSecret2.getBytes()).toEqual(sharedSecret.getBytes());
    });
    
    test('should throw error for invalid bytes length', () => {
      expect(() => {
        new SharedSecret(new Uint8Array(10));
      }).toThrow('Invalid shared secret length');
    });
    
    test('should throw error for invalid hex length', () => {
      expect(() => {
        SharedSecret.fromHex('abcd');
      }).toThrow('Invalid hex string length');
    });
  });
  
  describe('KEM operations', () => {
    test('should encapsulate and decapsulate correctly', () => {
      // Generate a key pair
      const keypair = keygen();
      
      // Encapsulate a shared secret
      const { ciphertext, sharedSecret: encapsulatedSecret } = encapsulate(keypair.publicKey);
      
      // Decapsulate the shared secret
      const decapsulatedSecret = decapsulate(keypair.privateKey, ciphertext);
      
      // Compare the shared secrets
      expect(decapsulatedSecret.getBytes()).toEqual(encapsulatedSecret.getBytes());
    });
  });
});